### ISSUE TITLE
Could pkg-fetch have looser version number matching when finding prebuilt binaries?

### ISSUE THREAD

**Issue by baparham** [vercel/pkg-fetch#283]
I have been thinking about this a bit, and looking for feedback.

Does `pkg-fetch` need to only find binaries that are explicitly versioned in the `patches.json` file when trying to download prebuilt binaries? My only thought here is that there is no `expected.ts` hash to compare against. What are some workarounds to this? Could we compare the published .sha256sum files?

Here's my thought: What if the latest release just contained any binary that we've managed to build from main, with the latest patches applied, and then a version of `pkg-fetch` (after this support is added, of course) could just ask the release for the latest version of the major node binary it's looking for? To verify the sha, it would need to download both the binary, and the .sha256sum file, compare then to the locally calculated sha256sum and then continue on, without needing to bump `pkg-fetch` versions to get the latest binaries within a reasonable time period?

any thoughts on this @robertsLando @leerob ... others?

Some questions:

What are the security implications to this approach?
What are the customer implications to stability? e.g. if users say `pkg --target node18` and they start getting 18.14.2 instead of 18.12.1 and running into issues like #282 ?
Could binary build and release be much more automated this way? e.g. the CI just needs to publish newer binaries to the existing release, which I'm 95% sure the github APIs support.

**Comment by robertsLando** [comment]
Don't speaking about security as it's not my field but from a user side I think this could have some issues. If a user really wants a new nodejs version he should bump pkg/pkg-fetch, otherwise two consecutive builds could result different without any clue why and then the reason could be we released a new patch and he was using a new node version that introduced a bug, this would be difficult to debug at first sight to users when running pkg

**Comment by baparham** [comment]
I suppose that use case could be guarded by creating a special node range qualifier, like `--target node18-latest` or something to indicate that it will always look for the latest version available, and users who are sensitive to version changes could do (like I do) `--target node18.14.2` but unless they actively choose to use `--target node18-latest` or similar they would get the existing behavior where the versions defined in the `patches.json` file are the only ones allowed

**Comment by robertsLando** [comment]
> I suppose that use case could be guarded by creating a special node range qualifier,

I like this, as long as there is back compatibility it's ok for me 👍🏼

**Comment by baparham** [comment]
I gave this some more thought I think this violates the basic security underpinning our strategy, which is to hard code the hashes to the source code to more clearly spot a compromise. Having the shas attached to the release is simply a false sense of security since maintainers are not necessarily within the trust boundary (though we try to be!), as was mentioned in #169 a compromise of any of the maintainers github accounts would allow for editing the release assets and content easily to produce malicious binaries with accompanying valid sha256sums. The only guard we have against this is the visibility into the source code changes for expected hashes that come from github actions with build steps we can audit.

I think this is mostly a workaround for having non-automated releases. Once we get something working that makes releasing smoother, I think the idea of upgrading patch versions to get new binaries is not so bad, we just have to be more regular about releasing new binaries.

**Comment by robertsLando** [comment]
> One option to allow a user to trust the contents of a release is to hard code the contents of a release into commits to the main branch, and run nightly tests against whether the shas on main match the shas in the release assets. If a failure occurs, we could perhaps quarantine any incorrect binaries or delete them from the release while we investigate what went wrong.

Not optimal as in this way there could be some hours with a malicious nodejs version in releases that could infect some users in the meantime that the workflow discovers it. Couldn't we simply do this and use the release [edited](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#release) event trigger instead of a cron? This way every time a release is edited we run it to verify nothing has been compromised and we are always covered

**Comment by baparham** [comment]
yeah, both on push to main, and on release edited, that would be the best

**Comment by robertsLando** [comment]
I think we could go with that so 👍🏼
